File: services/slackoauth/oauth.go
package slackoauth

import (
	"context"
	"encoding/json"
	"fmt"
	"io/ioutil"
	"net/http"
	"net/url"
	"time"

	"github.com/leighmcculloch/looks.wtf/shared/secrets"
	"google.golang.org/appengine/datastore"
	"google.golang.org/appengine/log"
	"google.golang.org/appengine/urlfetch"
)

type slackOauth struct {
	Ok          bool   `json:"ok"`
	AccessToken string `json:"access_token"`
	Scope       string `json:"scope"`
	UserID      string `json:"user_id"`
	TeamName    string `json:"team_name"`
	TeamID      string `json:"team_id"`
	Timestamp   time.Time
}

func persistSlackOauth(c context.Context, so slackOauth) error {
	so.Timestamp = time.Now().UTC()
	key := datastore.NewIncompleteKey(c, "slackOauth", nil)
	_, err := datastore.Put(c, key, &so)
	return err
}

func oauthHandler(c context.Context, w http.ResponseWriter, r *http.Request) error {
	code := r.URL.Query().Get("code")
	log.Infof(c, "Request with code: %#v", code)

	slackClientID := secrets.Get(c, "SLACK_CLIENT_ID")
	slackClientSecret := secrets.Get(c, "SLACK_CLIENT_SECRET")

	client := urlfetch.Client(c)
	resp, err := client.PostForm(
		"https://slack.com/api/oauth.access",
		url.Values{
			"client_id":     {slackClientID},
			"client_secret": {slackClientSecret},
			"code":          {code},
		},
	)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	body, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return err
	}

	var msg slackOauth
	err = json.Unmarshal(body, &msg)
	if err != nil || !msg.Ok {
		log.Errorf(c, "Error in response: %#v", string(body))
		return err
	}

	log.Infof(c, "Success, response: %#v", msg)
	fmt.Fprintf(w, "The looks.wtf Slack App has been successfully added!\nGive it a go with the `/look awe` command in Slack.")
	err = persistSlackOauth(c, msg)
	if err != nil {
		log.Infof(c, "Failed to persist: %#v", msg)
		return err
	}
	return nil
}

File: functions/oauth.js
export async function onRequest(context) {
  const code = new URL(context.request.url).searchParams.get('code');

  const response = await fetch('https://slack.com/api/oauth.access', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/x-www-form-urlencoded',
    },
    body: new URLSearchParams({
      client_id: context.env.SLACK_CLIENT_ID,
      client_secret: context.env.SLACK_CLIENT_SECRET,
      code: code,
    }).toString(),
  });

  const body = await response.json();
  if (body.ok) {
    console.log(`Ok, response: ${JSON.stringify(body)}`);
    return new Response('The looks.wtf Slack App has been successfully added!\nGive it a go with the `/look awe` command in Slack.');
  } else {
    console.log(`Not ok, response: ${JSON.stringify(body)}`);
    throw new Error('Not ok response from Slack');
  }
}
